### PR TITLE
[Dev] Add DeepEP v2 flex dispatcher backend

### PR DESCRIPTION
**Pull request by Autumn1998**
# What does this PR do ?
The DeepEP V2 support as a backend of flex dispatcher. 

PR on main: #5153 

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact the @mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/transformer/moe/fused_a2a.py [hunk 1]
```diff
@@ -270,6 +312,184 @@ def set_deepep_num_sms(num_sms):
...
+            grad_output.contiguous(),
+            handle=ctx.handle,
+            num_sms=ctx.num_sms,
+            previous_event=previous_event,
+            async_with_compute_stream=ctx.async_finish,
+            allocate_on_comm_stream=ctx.allocate_on_comm_stream,
+        )
+        if ctx.async_finish:
+            event.current_stream_wait()
+        return None, grad_x, None, None, None, None, None
```
Comment (Victarry) [inline megatron/core/transformer/moe/fused_a2a.py]: `DeepepV2Combine.forward` has six inputs after `ctx` (`buffer, x, handle, num_sms, async_finish, allocate_on_comm_stream`), but this backward returns seven gradients. PyTorch autograd will raise an incorrect-gradient-count error when this combine participates in training backward. Can we drop one trailing `None` here?

```python
return None, grad_x, None, None, None, None
```

File: megatron/core/transformer/moe/token_dispatcher.py [hunk 2]
```diff
@@ -1447,6 +1450,109 @@ def get_restored_hidden_states_by_experts(self, hidden_states: torch.Tensor) ->
...
+        self.group = group
+        self.num_local_experts = num_local_experts
+        self.config = config
+
+        self.router_topk = router_topk
+        self.num_experts = num_experts
+        self.router_dtype = config.moe_router_dtype
+        self.capacity_factor = config.moe_expert_capacity_factor
+        self.permute_fusion = config.moe_permute_fusion
+        self.num_sms = config.moe_deepep_num_sms
```
Comment (Victarry) [inline megatron/core/transformer/moe/token_dispatcher.py]: Question: should the v2 path leave `num_sms` as `0`, or use `buffer.get_theoretical_num_sms(...)`, instead of reusing `moe_deepep_num_sms`? DeepEP v2's `ElasticBuffer` can analytically choose SM/QP counts, so carrying over the v1 default may override the intended v2 behavior/perf tuning.

Comment (Autumn1998) [reply megatron/core/transformer/moe/token_dispatcher.py]: However, this also means we lose the args for adjusting the number of DeepEP v2 SMs. I prefer to keep this, but set the default value to None, and then let v1 and v2 use different default values when it is None.

File: megatron/core/transformer/moe/fused_a2a.py (hunk 1 above)
Comment (Autumn1998) [reply megatron/core/transformer/moe/fused_a2a.py]: fixed

**Review by Victarry — APPROVED** [review]
